### ISSUE TITLE
tree.Anonymous.genCSS does not stringify value and breaks source maps

### ISSUE THREAD

**Issue by rjgotten** [less/less.js#3022]
In Less 2.6.1 the `genCSS` method of `Anonymous` tree nodes does not take care to properly stringify its `value` when sent to `output.add`.

When source-maps are enabled, this results in an error `SyntaxError: chunk.split is not a function` because `SourceMapOutput` always expects its first `chunk` parameter to be a string on which it can call the `split` method.

From a cursory glance at the codebase, it seems Less 2.7.x still has this problem as well.

A minimal reproducing case:
```
.test { value : get-unit(10px); }
```

Proposed fix:

```js
Anonymous.prototype.genCSS = function (context, output) {
    var strValue = this.value != null ? String(this.value) : "";
    output.add(strValue, this.currentFileInfo, this.index, this.mapLines);
};
```

(Sorry guys; I don't have time to set up a full pull request. I hope someone can take this and do the legwork.)

**Comment by seven-phases-max** [comment]
>`.test { value : get-unit(10px); }`

I'm curious though how come `px`  is not a string here. Probably you mean `.test { value : get-unit(10); }`?

**Comment by matthew-dean** [comment]
@seven-phases-max A quoted string isn't necessary (or desirable) for values with units. See: http://lesscss.org/functions/#misc-functions-get-unit

**Comment by seven-phases-max** [comment]
@matthew-dean

>A quoted string isn't necessary (or desirable) for values with units.

No, I mean how come `px` value is not represented internally as JavaScript string.

I did some test and indeed it's `10px` casing the error. But the actual problem there seems to be `get-unit` itself returning `Anonymous` with value of `Dimension` type where `px` is numerator (i.e. it's a `get-unit` bug - not `Anonymous`).

**Comment by seven-phases-max** [comment]
So the minimal test to show `get-unit` bug:
```
.test {
    a: (3em * 2px / 1s);          // 6em
    b: get-unit(3em * 2px / 1s);  // em*px/s <- not expected here, should be em
}
```

**Comment by matthew-dean** [comment]
> No, I mean how come px value is not represented internally as JavaScript string.

Ah, that makes sense. It should definitely return an Anonymous string value, but it should be parsed as a value with a unit, I believe.

**Comment by rjgotten** [comment]
> @seven-phases-max 
> the actual problem there seems to be `get-unit` itself returning `Anonymous` with value of `Unit` type .

Hah; you're right! I missed that before.
In that case, the initial reported problem would be better fixed by having `get-unit` return just that `Unit` type node, instead of wrapping it into `Anonymous`. `Unit` correctly implements `compare` for comparisons in `Condition` nodes as used with mixin guards and it implements `genCSS` for CSS generation if you ever end up with Less that emits the unit into CSS.

It's just the semi-broken `toString` method then. At worst that would give you a wrong compilation result. But atleast not a full compiler crash, like happens now.

**Comment by seven-phases-max** [comment]
>In that case, the initial reported problem would be better fixed by having get-unit return just that Unit type node

I actually think it still return `Anonymous` (with its `value` always to be a js-string). The problem with `Unit` is that it's sort-of overengineered and the more we propagate it through the sources the more unexpected issues we met. So I believe that the `Unit` should be used *only* where its absolutely necessary (and imho, there's totally no benefit of returning it in `get-unit` - well, unless I miss something).

E.g. given the above example: there's no situation/context where `get-unit(3em * 2px / 1s);` returning anything but `Anonymous("px")` makes sense (so returning `Unit` with all of its never used `em*px/s` stuff is even harmful not just useless).

**Comment by matthew-dean** [comment]
> The problem with Unit is that it's sort-of overengineered

I think that's probably true for a number of Less.js nodes, where they are complete, distinct node types even though they could probably be a flatter list of nodes. Like, "10px" ends up being a `Dimension` node with a numerical value and a `Unit` node.

As I've started looking at documenting the Less API, I'm planning to omit a bunch of node types, because there's no way an end-user ever should have to know about many of them. A unit is the string part of a dimension. You're right that it's not as complicated as it needs to be.

**Comment by seven-phases-max** [comment]
Technically I'd say let's think of `Unit` as an *internal* type of `Dimension` that should never appear outside of it.

**Comment by matthew-dean** [comment]
@seven-phases-max Agreed. It should just be a string property on a `Dimension` (set to null for unit-less values), unless there's some unit comparison I'm not thinking of.

**Comment by rjgotten** [comment]
If there's no reason to keep `Unit` as a separately accessible node type; by all means, axe it.

The added utility of being able to run functions against `Unit` nodes is kind of moot, considering you could set those functions up to run against `Dimension` nodes as well, foregoing the need to have developers first extract the units with `get-unit`.

**Comment by matthew-dean** [comment]
@rjgotten 

If plugin authors could simply call `toString()` or `toCSS()` on these types of individual nodes that call nested nodes, it wouldn't be as much of an issue. But, as it is, flattening is such a chore that deeper nodes are hard to work with.
